### DISCUSSION TITLE
New libraries, stubs and autodetect approach

### DISCUSSION THREAD

**Discussion by elmot** [JetBrains/intellij-micropython#336] [Ideas]
**The problem:**

There are numerous Micropython boards nowadays, and a number of Micropython forks. There is a number of additional libraries to be installed. 
As a result, now it's virtually impossible to maintain a comprehensive set of python stubs inside the plugin. Additionally, that diversity of boards makes board detection by USB VID:PID useless in many cases.

**Proposed solution**

* Get rid of internal stubs completely.
* Change project setup approach: 
  1. User selects a com port among existing ones
  2. The plugin connects to the port and executes `os.uname()`
  3. Based on the gathered information, the plugin recommends to install a best suitable stub(s) from PyPl repository
  4. The user confirms the choice 
  5. Stubs are installed using `pip` into the project structure
  6. The user is able to install any source or compiled libraries to the project, and their stubs to a dedicated folder.
  
<details>
  <summary>As the result, a typical project layout with `main.py` and compiled `ssd1306` library would look like: </summary>
 
 ![image](https://github.com/user-attachments/assets/62f1859e-f5db-4807-b8ce-e5e48f2ab19c)
 
 Device layout looks like 
```
 main.py
 lib/ssd1306.mpy
```
</details>
  
Any thoughts or concerns?

Related to #220, #240, #141, #114, #326, #319, #316, and many others

**Comment by lukaskremla** [comment]
I may have misunderstood, but judging by the image having the stubs folders marked as source root, I got the impression that the proposed solution entails uploading stubs directly onto the device filesystem. If that is the case, I would avoid doing so. Microcontrollers are by definitions working with restricted resources, and any amount of free storage space is valuable, this is especially true for some ESP8266 devices. 

If I have indeed misunderstood or if there is a way to do what you propose without uploading stubs on the device file system, then I believe that this is actually the best practice - involving the user in selecting the stub device type. 

There are a few edge case scenarios to handle that come to mind. The plugin probably should allow the user to somehow manually select/search for the specific device type, if for whatever reason the detection fails or the user has some kind of alternative intent. This could be handled together with manually selecting one's own local stubs.

Aside from uploading stubs directly onto the device, which might just be my misunderstanding, this solution seems like the best way to handle stubs in this plugin.

**Reply by elmot** [reply]
Of course, no stubs are going to be uploaded. I have not decided yet how to select uploadable parts.
Easiest idea is that all the files names with leading dot will be skipped. 
So, `.gitignore`, as well as various `.venv`, `.xxxx-stub` etc won't be uploaded

<details>
<summary>Two possible layouts:</summary>

![image](https://github.com/user-attachments/assets/d5b6245b-3edf-4276-97d4-05216b0475f4)
or
![image](https://github.com/user-attachments/assets/bd4b4028-11ee-40eb-b1ea-09200654b854)

</details>

**Reply by lukaskremla** [reply]
> Of course, no stubs are going to be uploaded. I have not decided yet how to select uploadable parts. Easiest idea is that all the files names with leading dot will be skipped. So, `.gitignore`, as well as various `.venv`, `.xxxx-stub` etc won't be uploaded
> 
> Two possible layouts:
> ![image](https://private-user-images.githubusercontent.com/5366945/385217338-d5b6245b-3edf-4276-97d4-05216b0475f4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0NDQ3NTEsIm5iZiI6MTczMTQ0NDQ1MSwicGF0aCI6Ii81MzY2OTQ1LzM4NTIxNzMzOC1kNWI2MjQ1Yi0zZWRmLTQyNzYtOTdkNC0wNTIxNmIwNDc1ZjQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTJUMjA0NzMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTNhYjlmNTVmNmU4NDg1ODczMWVlZGU2ZDJlZTM1Y2RiNWNjMTM2YmJkYWI1YmNlMzA5MDU5MGEyOGJhMDE4NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.nMmS7HOWNCmH9XofWOabQzA_lPPdByjNTKcjRRp_TFU) or ![image](https://private-user-images.githubusercontent.com/5366945/385218838-bd4b4028-11ee-40eb-b1ea-09200654b854.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0NDQ3NTEsIm5iZiI6MTczMTQ0NDQ1MSwicGF0aCI6Ii81MzY2OTQ1LzM4NTIxODgzOC1iZDRiNDAyOC0xMWVlLTQwZWItYjFlYS0wOTIwMDY1NGI4NTQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTJUMjA0NzMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjY3NjRlYzI2ODdkMjJmMGE2ZGRkZDUxODQ5MGFiYmRiNzliZDExYjQzNzljMzk1MDI0M2UzMmM2ODZhZGZmMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.BaBRPhxOfa4GgeX81QyamNbcwtgy3bQL9Gg5sI6c-88)

To me using the mark directory functionality for determining what gets uploaded to my device quite intuitive. The project I work on has two folders marked as source folders to make things more readable and organized, and with how the plugin treats them their contents get uploaded to the device file system without the containing folders, allowing me to organize both the Project and File System layout more freely. 

If there was just a single pre-generated folder that would be meant to store all files and folders that are meant to be uploaded, it would give users less freedom as far as organizing their projects go.

Maybe it's sufficient to only exclude files like `.gitingore` that start with a dot and keep relying on "mark as excluded" for folders. Though I don't really see a reason I would want to use a leading dot to name a folder which should be uploaded.

**Reply by elmot** [reply]
Excluded folder can't contain any stubs. The folder is completely ignored, and the language engine just ignores it.

Having several source folders is totally fine.
My idea is that all the folders marked "Source" are taken into account by the language engine to analyze sources and making all the features possible.
Source folders started with dot are analyzed, but does not go to the device.
Normal source folders content goes to the device without folder names(1st proposed layout)
If there is no source folders without dots, then the project root goes to the device excluding leading-dot files(2nd proposed layout)

**Reply by lukaskremla** [reply]
> Excluded folder can't contain any stubs. The folder is completely ignored, and the language engine just ignores it.
> 
> Having several source folders is totally fine. My idea is that all the folders marked "Source" are taken into account by the language engine to analyze sources and making all the features possible. Source folders started with dot are analyzed, but do not go to the device. Normal source folders content goes to the device without folder names(1st proposed layout). If there is no source folders without dots, then the project root goes to the device excluding leading-dot files(2nd proposed layout)

I understand now. What you propose makes sense, I can't think of a better way to handle this.

**Reply by elmot** [reply]
Formalized in #338

**Reply by lukaskremla** [reply]
> Formalized in #338

I had an idea for adding more upload options while testing the pre-alpha 12 release. Currently you have the option to "Upload File/Directory to MP device". Well, for directories you could have more options such as upload entire file/upload contents. I can, however, see how this might be overcomplicating things with how the current uploading priority system with source folders works. I just thought I'd leave this here as an idea in case it could be useful to you somehow.

**Comment by lukaskremla** [comment]
There is an alternative approach to auto-detection compared to the old one.

I personally go back and forth between multiple ESP32 devices and USB ports when working with MicroPython. Different USB hub at work, different USB hub at home and even switching to a different USB port on my computer changes the device path of the Microcontroller I am working with.

Perhaps a different form of auto-detection could be implemented in the new plugin. It would make things a tad bit easier if there was a way to track one device even after it pops up under a different device path. Known devices could be saved as device presets, or just the last connected device could be remembered and searched for in the available device paths. It doesn't really matter how complex the functionality built around this device tracking would be.

Something like this could theoretically even be used to narrow down the USB serial port selection drop-down list to just devices that can somehow be verified as having MicroPython installed. 

Now implementing this is where things can get more difficult. Telling apart devices which have MicroPython installed [is possible](https://docs.micropython.org/en/latest/library/machine.html#machine.unique_id). The unique ID can be used and relied upon on most if not all device types. So some combination of the USB VID:PID and unique ID could be used. I can, however, see how scanning all serial port paths might not be practical. 

I will leave the question of whether or not this is plausible up to you, you are better suited to answer it than me. But talking about autodetect approach this is what comes to mind for me.

If this would indeed be possible and wouldn't require a terribly overcomplicated solution, then some really cool features could be built around it.

**Reply by elmot** [reply]
How many uart ports do you have visible simultaneously? Max and average?

**Reply by lukaskremla** [reply]
> How many uart ports do you have visible simultaneously? Max and average?

At home it can be 2-3, and at work it is similar most of the time, max could be 5, but that is very rare.

**Reply by Josverl** [reply]
I have reports of users having in excess 32 serial ports connected in their setup and have needed to adapt mpflash for that. 

It is possible to interrogate the micropython port/boart/variant and version of connected devices , but it does interrupt any running code.
[`mpflash list --json`](https://pypi.org/project/mpflash/)

**Reply by elmot** [reply]
That's a lot...

**Reply by Josverl** [reply]
This is the MicroPython code that mpflash uses to get the info. Tested from  v1.9.3 ... v1.23.0

https://github.com/Josverl/micropython-stubber/blob/main/src%2Fmpflash%2Fmpflash%2Fmpremoteboard%2Fmpy_fw_info.py

**Reply by elmot** [reply]
I'd rather use mpflash as is.

**Comment by Josverl** [comment]
If there are plans to take this approach , 
I would welcome 
- adopting the current JetBrains stubs if they can be released onder MIT
- merge them with the current micropython-stubs 
- update them to the recent micropython ( v1.23.x and v1.24.x )
- publish the combined result to PyPi 
- add them to the planned integration of stubs into the main micropython repo (already underway)   

I have a preference for MIT, mostly to avoid carrying too many different license types into the micropython repo.

If that is not possible, then most likely I'll go back to the the somewhat older version from https://github.com/hlovatt/PyBoardTypeshed instead. But before that we need to discuss that with the MicroPython core team as well.
I have no view on the incurred Delta difference between these 

Also see: https://github.com/Josverl/micropython-stubs/issues/762

**Reply by elmot** [reply]
Oh, you are very welcome to the discussion! 
(As far as I understand, you are the stubs repository maintainer)
Is there any value to merge our stubs to your ones? Our ones are quite outdated.

And I a have a feature request for you - list available stubs somehow. Either by loading the list from some known url or by calling mpremote with a special key.
I also have similar request for packages in micropython-lib, but that is another story.

**Reply by Josverl** [reply]
These ones have @overloads defined, which i did not yet add, 
So that  helps connect the dots, hopefully.


What do you need for a list of stubs ?
I probably have the data hidden away somewhere

**Reply by elmot** [reply]
>What do you need for a list of stubs 

something like
`mpflash --list-stubs [--port rp2|esp32|...] [--version micropython_version]`

or it might be `mpremote mip ...`, whatever works better

> adopting the current JetBrains stubs if they can be released under MIT

The license holder is JetBrains, so we need to ask our lawyers first

**Reply by Josverl** [reply]
Understood, I'm also not a lawyer
- I'll start with @hlovatt stubs, as they are MIT licensed. 

For the stub identification / mpflash suggestion 
- I think that could work by looking up the port/board/variant against a cached list of  the most recent version published stubs.
so something like :
`mpflash --list --json` 
```json 
[
    {
        "serialport": "COM14",
        "connected": true,
        "family": "micropython",
        "description": "Generic ESP32 module with ESP32",
        "version": "1.23.0",
        "port": "esp32",
        "board": "ESP32_GENERIC",
        "cpu": "ESP32",
        "arch": "xtensawin",
        "mpy": "v6.3",
        "build": "",
        "location": "1-6.1",
        "stub-package": "micropython-esp32-stubs==1.23.0.*",
         "toml": {}
    }
]
```

Would that work , or would you also want an option to get a list of all published stubs ?
Probably simplest if I use a GH Action to publish that on github as a json or csv.
```csv
name,mpy_version
micropython-esp32-stubs,1.23.0-preview
micropython-samd-stubs,1.23.0-preview
micropython-esp32-esp32_generic-stubs,1.23.0
micropython-esp32-esp32_generic_c3-stubs,1.23.0
micropython-esp32-esp32_generic_s3-stubs,1.23.0
micropython-esp32-stubs,1.23.0
micropython-esp8266-esp8266_generic-stubs,1.23.0
micropython-esp8266-stubs,1.23.0
micropython-rp2-rpi_pico-stubs,1.23.0
micropython-rp2-rpi_pico_w-stubs,1.23.0
micropython-rp2-stubs,1.23.0
micropython-samd-seeed_wio_terminal-stubs,1.23.0
micropython-samd-stubs,1.23.0
micropython-stm32-pybv11-stubs,1.23.0
micropython-stm32-stubs,1.23.0
micropython-rp2-arduino_nano_rp2040_connect-stubs,1.22.1
micropython-rp2-rpi_pico-stubs,1.22.1
micropython-rp2-rpi_pico_w-stubs,1.22.1
micropython-rp2-stubs,1.22.1
micropython-samd-seeed_wio_terminal-stubs,1.22.1
micropython-samd-stubs,1.22.1
micropython-esp32-esp32_generic-stubs,1.22.0
micropython-esp32-esp32_generic_s3-stubs,1.22.0
micropython-esp32-stubs,1.22.0
micropython-rp2-arduino_nano_rp2040_connect-stubs,1.22.0
micropython-rp2-pimoroni_picolipo_16mb-stubs,1.22.0
micropython-rp2-rpi_pico-stubs,1.22.0
micropython-rp2-rpi_pico_w-stubs,1.22.0
micropython-rp2-stubs,1.22.0
micropython-samd-seeed_wio_terminal-stubs,1.22.0
micropython-samd-stubs,1.22.0
micropython-stm32-pybv11-stubs,1.22.0
micropython-stm32-stubs,1.22.0
micropython-esp32-esp32_generic-stubs,1.21.0
micropython-esp32-stubs,1.21.0
micropython-rp2-rpi_pico-stubs,1.21.0
micropython-rp2-rpi_pico_w-stubs,1.21.0
micropython-rp2-stubs,1.21.0
micropython-samd-seeed_wio_terminal-stubs,1.21.0
micropython-samd-stubs,1.21.0
micropython-stm32-pybv11-stubs,1.21.0
micropython-stm32-stubs,1.21.0
micropython-unix-stubs,1.21.0
micropython-webassembly-stubs,1.21.0
micropython-esp32-generic_ota-stubs,1.20.0
micropython-esp32-generic_s3-stubs,1.20.0
micropython-esp32-ota-stubs,1.20.0
micropython-esp32-s3-stubs,1.20.0
micropython-esp32-stubs,1.20.0
micropython-rp2-pico-stubs,1.20.0
micropython-rp2-pico_w-stubs,1.20.0
micropython-rp2-pimoroni_picolipo_16mb-stubs,1.20.0
micropython-rp2-stubs,1.20.0
micropython-samd-adafruit_feather_m4_express-stubs,1.20.0
micropython-samd-adafruit_itsybitsy_m4_express-stubs,1.20.0
micropython-samd-minisam_m4-stubs,1.20.0
micropython-samd-seeed_wio_terminal-stubs,1.20.0
micropython-samd-stubs,1.20.0
micropython-stm32-pybv11-stubs,1.20.0
micropython-stm32-stubs,1.20.0
micropython-unix-stubs,1.20.0
micropython-esp32-generic_s3-stubs,1.19.1
micropython-esp32-generic_spiram-stubs,1.19.1
micropython-esp32-s3-stubs,1.19.1
micropython-esp32-stubs,1.19.1
micropython-esp32-um_tinypico-stubs,1.19.1
micropython-esp8266-stubs,1.19.1
micropython-rp2-stubs,1.19.1
micropython-stm32-stubs,1.19.1
micropython-esp32-stubs,1.18
micropython-esp8266-stubs,1.18
micropython-rp2-stubs,1.18
micropython-stm32-stubs,1.18
micropython-esp32-stubs,1.17
micropython-esp8266-stubs,1.17
micropython-rp2-stubs,1.17
micropython-stm32-stubs,1.17
```

**Reply by Josverl** [reply]
Short update: 
I was able to create a quite decent set of [reference stubs](https://github.com/Josverl/micropython-stubs/tree/main/micropython-reference), 
based on only MIT sources, so I don't think there is much additional information in the Jetbrain stubs.

But there is additional work to be done to to get that percolated down to the individual port and board stub-packages, so ill continue down that path.

**Reply by elmot** [reply]
> Would that work , or would you also want an option to get a list of all published stubs ?
Probably simplest if I use a GH Action to publish that on github as a json or csv.

Full list is probably better.

>I was able to create a quite decent set of [reference stubs](https://github.com/Josverl/micropython-stubs/tree/main/micropython-reference),

Good job!

>  I don't think there is much additional information in the Jetbrain stubs.

Does it mean you do not need them under MIT anymore?

**Reply by Josverl** [reply]
- Full list in JSON is located here 
  https://github.com/Josverl/micropython-stubs/blob/main/data/stub-packages.json 
  - I will need to integrate this in CI/CD , but first let me know if this format works for you
  
- MIT license - not lets save ourselves the hassle of needing to involve legal - there will need to be manual adjustments and verification anyway

**Comment by lukaskremla** [comment]
Hey @elmot could I ask about the reason you want to switch to using mpremote? What are the benefits over the existing Kotlin implementation?

**Reply by elmot** [reply]
Sure.
mpremote pros:
* we can get rid of responsibility to maintain communication level 
* we can get rid of responsibility to maintain file exchange protocols
* mpremote guys promise to support crc checks on the library level, that optimizes uploads
* ability to support `mpremote mip` to install libraries

Concerns:
 * need to switch back from pure Kotlin to to python process interaction
 * webrepl is not supported yet

**Reply by lukaskremla** [reply]
> Sure. mpremote pros:
> 
> * we can get rid of responsibility to maintain communication level
> * we can get rid of responsibility to maintain file exchange protocols
> * mpremote guys promise to support crc checks on the library level, that optimizes uploads
> * ability to support `mpremote mip` to install libraries
> 
> Concerns:
> 
> * need to switch back from pure Kotlin to to python process interaction
> * webrepl is not supported yet

> Sure. mpremote pros:
> 
> * we can get rid of responsibility to maintain communication level
> * we can get rid of responsibility to maintain file exchange protocols
> * mpremote guys promise to support crc checks on the library level, that optimizes uploads
> * ability to support `mpremote mip` to install libraries
> 
> Concerns:
> 
> * need to switch back from pure Kotlin to to python process interaction
> * webrepl is not supported yet

I've looked into mpremote and its capabilites, I do understand why you want to switch, but I am not sure I'd dwell on using mpremote right now or supporting WebREPL right from the start.

Some thoughts about this:

1. Mpremote's hash checks aren't optimal performance wise - they use SHA256 calculations that happen each time before a file is uploaded. First mpremote attempts to calculate the hash on device, if that fails the file is downloaded and the hash is calculated and compared locally. This wastes a lot of time and the fallback local calculation seems like an entirely wasted step in this plugin's use case.
2. Mpremote does not maintain a permanent connection like this plugin does. Each time you want to execute a command in it you have to run mpremote connect <device> <command>, this might be problematic given how this plugin is structured.
3. WebREPL is indeed not supported currently and there is no concrete date on when it might be. There's no working MicroPython plugin for PyCharm or other major IDEs that is universal and works decently, a least I'm not aware of one. I don't think that the development of this plugin should get bogged down on waiting for mpremote WebREPL support.
4. I assume you want to integrate the tool as is without modifying its source code (given you want to avoid the comms layer maintenance) - could the wrapper or whatever will need to be made to interact with mpremote slow things down? 

Point 1. could be circumvented by creating a custom hashing implementation and skipping the built-in one for every upload. I am working on such script already, I could provide it if necessary.

I really think it would be ideal for the community to ship something out as soon as possible without waiting for mpremote to support WebREPL.  

There is a huge gap in the market of MicroPython IDE plugins. The old MicroPython plugin for Jetbrains is no longer even supported and available for download in the Jetbrains plugin marketplace and even when it was available, there was a ton of bugs in it that made it an incredible hassle to work with. There's a ton of developers out there who would greatly appreciate this plugin even in the form in which it currently is if the major bugs (ones I made pull request for) are solved.

**Reply by elmot** [reply]
Thanks for the review, that is very interesting. First of all, there is no decision yet, present pure-kotlin communication works somehow, I just don't want to touch it for a while, and then either switch to mpremote or implement various optimizations in kotlin. None of that is gonna be done before nova release. => No delays, no 

>Mpremote's hash checks aren't optimal performance wise - they use SHA256 calculations that happen each time before a file is uploaded. First mpremote attempts to calculate the hash on device, if that fails the file is downloaded and the hash is calculated and compared locally. This wastes a lot of time and the fallback local calculation seems like an entirely wasted step in this plugin's use case.

Sounds odd to me. The only advantage(and quite questionable one) is this approach do decrease flash wears. If I implement this, I'd rather unconditionally upload/download the file if the hashing fails. 
Probably it's better if you talk to mpremote guys or even contribute your code to the library. They communicate pretty well on discord.

> Mpremote does not maintain a permanent connection like this plugin does.

I know, but I have an idea how to deal with that.

>WebREPL is indeed not supported currently and there is no concrete date on when it might be.

Sure. Let's look at it again when there is the moment for optimizations

>I assume you want to integrate the tool as is without modifying its source code (given you want to avoid the comms layer maintenance) - could the wrapper or whatever will need to be made to interact with mpremote slow things down?

I don't think so. The bottleneck no 1 is always communication level, the bottleneck #2 is flash write speeds. The python interaction works on 3Ghz CPUs and takes almost nothing.

> Point 1. could be circumvented by creating a custom hashing implementation and skipping the built-in one for every upload. I am working on such script already, I could provide it if necessary.

Unfortunately, that means the custom hashing code is not a part of the mpy library on the device, and it is uploaded there every time, and it goes through the same bottleneck, and slows down everything, thus mitigates the optimizations

**Reply by lukaskremla** [reply]
> Thanks for the review, that is very interesting. First of all, there is no decision yet, present pure-kotlin communication works somehow, I just don't want to touch it for a while, and then either switch to mpremote or implement various optimizations in kotlin. None of that is gonna be done before nova release. => No delays, no
> 
> > Mpremote's hash checks aren't optimal performance wise - they use SHA256 calculations that happen each time before a file is uploaded. First mpremote attempts to calculate the hash on device, if that fails the file is downloaded and the hash is calculated and compared locally. This wastes a lot of time and the fallback local calculation seems like an entirely wasted step in this plugin's use case.
> 
> Sounds odd to me. The only advantage(and quite questionable one) is this approach do decrease flash wears. If I implement this, I'd rather unconditionally upload/download the file if the hashing fails. Probably it's better if you talk to mpremote guys or even contribute your code to the library. They communicate pretty well on discord.
> 
> > Mpremote does not maintain a permanent connection like this plugin does.
> 
> I know, but I have an idea how to deal with that.
> 
> > WebREPL is indeed not supported currently and there is no concrete date on when it might be.
> 
> Sure. Let's look at it again when there is the moment for optimizations
> 
> > I assume you want to integrate the tool as is without modifying its source code (given you want to avoid the comms layer maintenance) - could the wrapper or whatever will need to be made to interact with mpremote slow things down?
> 
> I don't think so. The bottleneck no 1 is always communication level, the bottleneck #2 is flash write speeds. The python interaction works on 3Ghz CPUs and takes almost nothing.
> 
> > Point 1. could be circumvented by creating a custom hashing implementation and skipping the built-in one for every upload. I am working on such script already, I could provide it if necessary.
> 
> Unfortunately, that means the custom hashing code is not a part of the mpy library on the device, and it is uploaded there every time, and it goes through the same bottleneck, and slows down everything, thus mitigates the optimizations

I've thought of a different approach, which shaves off an incredible amount of time and avoids interrogating the device for a hash for every single file uplaod - I already have it integrated in my version of this plugin and it's been tested by both me and my colleagues. Currently it uses SHA256 for hash comparisons, but my plan is to use FNV1A32 which I think should be pretty easy to implement in a few lines - meaning it should run on any micropython port. I can create a pull request of this script and another upload related functionality that I have created if you'd be interested. 

It reads in chunks into a bytearray and it has handling for cases where the necessary libraries aren't available - it will simply upload everything as if no matching files were found.

And regarding not touching the existing Kotlin implementation - I would like you to actually reconsider this, as there are existing bugs in the communication layer that are quite annoying - they can be mitigated by simply retrying, but there are several errors in the serial communication.

Sometimes the plugin times out on "import os,errno" command, which is a part of your blindExecute method, similarly sometimes the other upload commands in this method time out as well. And finally, there seems to be some issue with frequent doBlindExecute call. 

Back when I submitted my very basic CRC32 hash comparison method - I though there must be something wrong with it, because it caused super frequent timeout errors with this plugin, but it seems to me that there is actually some problem with calling blindExecute very frequently. 

When I made the CRC hash calculation script run before file uploads on ESP32 S3 devices, it would freeze and time out in 3-8 uploaded files.

My knowledge of Kotlin and its serial communication layer means I can't really contribute a fix to the serial communication problem - I'd like to learn about it, but currently I feel like I can contribute more by focusing on other things.

Regarding my custom script - I am still improving it and I do plan on discussing a potential merge into MicroPython with the devs, but I first want to finish the FNV1A32 implementation and run some experiments for using very large chunks safely by utilizing bytearrays.

**Reply by elmot** [reply]
There is already a crc32 calculator at mpy side, as far as I understand it native code, so should work really fast. It looks like it may work in by chunk, see [binascii.crc32(data[, value])](https://docs.micropython.org/en/latest/library/binascii.html#binascii.crc32)
That `data` parameter, which means ability to work by chunks, was either documented recently or I was just blind before.

**Reply by lukaskremla** [reply]
> There is already a crc32 calculator at mpy side, as far as I understand it native code, so should work really fast. It looks like it may work in by chunk, see [binascii.crc32(data[, value])](https://docs.micropython.org/en/latest/library/binascii.html#binascii.crc32) That `data` parameter, which means ability to work by chunks, was either documented recently or I was just blind before.

The problem is that this method is not included in the ESP8266 port, at least from what I understand - hence why I want to implement FNV1A32

**Reply by mattytrentini** [reply]
Yes, [CRC32 is disabled on the ESP8266](https://github.com/micropython/micropython/blob/master/ports/esp8266/mpconfigport.h#L41) (presumably to conserve the limited flash space). 

Note that Brian Pugh already has an [FNV1A32 implementation written as a native module](https://github.com/BrianPugh/micropython-fnv1a32)...though it sounds like you're after a pure Python implementation?

**Reply by lukaskremla** [reply]
> Yes, [CRC32 is disabled on the ESP8266](https://github.com/micropython/micropython/blob/master/ports/esp8266/mpconfigport.h#L41) (presumably to conserve the limited flash space).
> 
> Note that Brian Pugh already has an [FNV1A32 implementation written as a native module](https://github.com/BrianPugh/micropython-fnv1a32)...though it sounds like you're after a pure Python implementation?

Brian Pugh is actually the one who gave me this idea in the MicroPython discord and yes, I'm after a pure Python implementation.

**Comment by lukaskremla** [comment]
In the last few days I've been considering how stubs and device type detection could be handled. Firstly, I would welcome moving the port selector drop down left of the connect/disconnect button in the MicroPython plugin window. This would make it easier to work with, as I often find myself having to go to the settings multiple times in a day, which can get annoying.

I understand that this could be difficult and that it has plenty of considerations, here are a few problems and their solutions I thought of:

1. Users changing the serial port when connected - upon successfully connecting it would be great to gray out the dropdown to not be clickable. From my experience working with the Jetbrains UI API perhaps this could be possible with some disable parameter/property.
2. Refreshing the serial port drop down - would it be possible to refresh the serial port listing each time the drop down is expanded?
3. Filtering out irrelevant serial devices - here I would look at how Thonny does it, I believe, using USB VID and PID. Thonny's port selector looks sort of like this: `MicroPython (ESP32):  Espressif Device @ /dev/cu.usbmodem101`. Now this plugin could simplify it to `ESP32: /dev/cu.usbmodem101`. This would serve the purpose of making the selection clearer and it could also be used for handling stubs - although the best way here might be interrogating the device for specific device type and MicroPython version each time a connection is established.
4. There are two more considerations -> handling WebREPL and stubs when no device is connected:

For both of these problems I would go with an approach similar to how thonny does it. In Thonny you have the bottom right serial port drop down, which is the default quick access. Then its interpreter settings allow you to select a device manufacturer, specific communication type and then either the serial device or WebREPL credentials. The way this then propagates is that the main quick access serial device drop down is switched to displaying `MicroPython (ESP32): ws://192.168...` again this could be adapted like I suggested before. 

This plugin's implementation could be expanded for choosing the MicroPython version as well, so that users have more control over specific stub settings should they need it.

**Reply by elmot** [reply]
> In the last few days I've been considering how stubs and device type detection could be handled. Firstly, I would welcome moving the port selector drop down left of the connect/disconnect button in the MicroPython plugin window. This would make it easier to work with, as I often find myself having to go to the settings multiple times in a day, which can get annoying.
> 
> I understand that this could be difficult and that it has plenty of considerations, here are a few problems and their solutions I thought of:
> 
> 1. Users changing the serial port when connected - upon successfully connecting it would be great to gray out the dropdown to not be clickable. From my experience working with the Jetbrains UI API perhaps this could be possible with some disable parameter/property.
> 2. Refreshing the serial port drop down - would it be possible to refresh the serial port listing each time the drop down is expanded?
> 3. Filtering out irrelevant serial devices - here I would look at how Thonny does it, I believe, using USB VID and PID. Thonny's port selector looks sort of like this: `MicroPython (ESP32):  Espressif Device @ /dev/cu.usbmodem101`. Now this plugin could simplify it to `ESP32: /dev/cu.usbmodem101`. This would serve the purpose of making the selection clearer and it could also be used for handling stubs - although the best way here might be interrogating the device for specific device type and MicroPython version each time a connection is established.
> 4. There are two more considerations -> handling WebREPL and stubs when no device is connected:
> 
> For both of these problems I would go with an approach similar to how thonny does it. In Thonny you have the bottom right serial port drop down, which is the default quick access. Then its interpreter settings allow you to select a device manufacturer, specific communication type and then either the serial device or WebREPL credentials. The way this then propagates is that the main quick access serial device drop down is switched to displaying `MicroPython (ESP32): ws://192.168...` again this could be adapted like I suggested before.
> 
> This plugin's implementation could be expanded for choosing the MicroPython version as well, so that users have more control over specific stub settings should they need it.


> In the last few days I've been considering how stubs and device type detection could be handled. Firstly, I would welcome moving the port selector drop down left of the connect/disconnect button in the MicroPython plugin window. This would make it easier to work with, as I often find myself having to go to the settings multiple times in a day, which can get annoying.
> 
> I understand that this could be difficult and that it has plenty of considerations, here are a few problems and their solutions I thought of:
> 
> 1. Users changing the serial port when connected - upon successfully connecting it would be great to gray out the dropdown to not be clickable. From my experience working with the Jetbrains UI API perhaps this could be possible with some disable parameter/property.
> 2. Refreshing the serial port drop down - would it be possible to refresh the serial port listing each time the drop down is expanded?
> 3. Filtering out irrelevant serial devices - here I would look at how Thonny does it, I believe, using USB VID and PID. Thonny's port selector looks sort of like this: `MicroPython (ESP32):  Espressif Device @ /dev/cu.usbmodem101`. Now this plugin could simplify it to `ESP32: /dev/cu.usbmodem101`. This would serve the purpose of making the selection clearer and it could also be used for handling stubs - although the best way here might be interrogating the device for specific device type and MicroPython version each time a connection is established.
> 4. There are two more considerations -> handling WebREPL and stubs when no device is connected:
> 
> For both of these problems I would go with an approach similar to how thonny does it. In Thonny you have the bottom right serial port drop down, which is the default quick access. Then its interpreter settings allow you to select a device manufacturer, specific communication type and then either the serial device or WebREPL credentials. The way this then propagates is that the main quick access serial device drop down is switched to displaying `MicroPython (ESP32): ws://192.168...` again this could be adapted like I suggested before.
> 
> This plugin's implementation could be expanded for choosing the MicroPython version as well, so that users have more control over specific stub settings should they need it.

That is a big topic, let's move that to another discussion.

**Comment by lukaskremla** [comment]
I have a question regarding two functions that could be added to the plugin.

1. Board RTC synchronization - Thonny does this already, I plan on modifying the plugin to handle this somehow as well, should I create a PR when I'm done?
2. Synchronize run configuration option - This is related to the hashing script I've mentioned in a previous reply here. The script I have handles both the file hash comparisons and deleting files that are not a part of the upload, should I open a PR to discuss this more specifically as well?

**Reply by elmot** [reply]
> 1. Board RTC synchronization - Thonny does this already, I plan on modifying the plugin to handle this somehow as well, should I create a PR when I'm done?

Sure, you are welcome. That's useful addition

> 2. Synchronize run configuration option - This is related to the hashing script I've mentioned in a previous reply here. The script I have handles both the file hash comparisons and deleting files that are not a part of the upload, should I open a PR to discuss this more specifically as well?

Sure, but better to start with a discussion

**Comment by Josverl** [comment]
Just FYI ,

I released an update to the [MicroPython stubs for v1.24.1](https://github.com/Josverl/micropython-stubs/releases/tag/v1.24.1-stubs)
where the stubs now have similar information on parameters, return types and overloads.

That should bring static type checking and linting  on par or better than the stubs previously used by the intelli-j-micropython plugin
